### ISSUE TITLE
Database Evolutions - Integrating with Slick 2 Code Generator

### ISSUE THREAD

**Issue by papauschek** [playframework/play-slick#118]
I recently wrote an article about how to integrate **Play Framework Evolutions** with the new **Slick 2.0 code generator**, which enables you to write your evolutions by hand and get the necessary slick code autogenerated.

It's basically the reverse of what play-slick is doing right now, as I understand it (code-first via EBean).

It would be nice to have this functionality all in one place, and play-slick seems the perfect place for it, since I saw you already have a branch for upgrading to Slick 2.0

I have a working prototype for the code generator which integrates with evolutions right here:
https://github.com/papauschek/play-slick-evolutions
And here are some explanations:
http://blog.papauschek.com/2013/12/play-framework-evolutions-slick-2-0-code-generator/

My question is: can we somehow integrate this into play-slick? Right now, my prototype uses a separate play module for code generation. How could this be integrated in a Play module?

If someone with more SBT knowledge could point me to the right direction, I would be happy to help with the integration.

Looking forward to some feedback!

**Comment by nafg** [comment]
@cvogt @jroper Can't play evolutions be made to run scala (or java) code?

**Comment by cvogt** [comment]
no idea about Play evolutions but what you are asking for is basically what I prototyped here: https://github.com/cvogt/migrations

Never had the time to flesh it out. Maybe you want to take a stab at it :)?

**Comment by pathikrit** [comment]
Yes, take a a look at this:
https://github.com/pathikrit/garfield/blob/master/play-slick-evolutions-codegen-plugin/app/PlaySlickCodeGenerator.scala

I am trying to create an sbt plugin for this myself:
https://github.com/pathikrit/garfield/tree/master/play-slick-evolutions-codegen-plugin

**Comment by hepin1989** [comment]
should provide play-slick-codegen

**Comment by krivachy** [comment]
@cvogt What is the status with this? Has someone created a solution that works? I think this would be an awesome and much needed feature and could help implement it.

**Comment by cvogt** [comment]
no progress on this from my side particular to this ticket. However, since then "Scala Forklift" was developed https://github.com/lastland/scala-forklift . @lastland maybe you want to take a look how it covers this use case and how it would be integrated with slick to replace the built-in evolutions including slick support.

**Comment by lastland** [comment]
@cvogt thanks for letting me know! I will take a look at it!

**Comment by lastland** [comment]
Hi everyone,

I have just made an example of using Play, Play-Slick with Scala Forklift. You can find the example here: https://github.com/lastland/play-slick-forklift-example

Scala Forklift supports migration files written in plain SQL queries or type-safe Slick queries. It applies all the migrations one by one, and generates the corresponding schema code for each version automatically. It also supports recovering from previous break migration (for example, when the generated code cannot compile).

There is also an experimental feature to support developing with git branches. It is not yet demonstrated in the example, but you can find more information here: https://github.com/lastland/scala-forklift/tree/develop/example/tools/git

Basically you don't need much extra work to start a web application with Play, Play-Slick, and Forklift. Just clone the Forklift start template from https://github.com/lastland/scala-forklift-start-template, put the web application written with Play under the `app` directory, and edit some files accordingly (for example, edit [`migrations/src/main/scala/Migrations.scala`](https://github.com/lastland/play-slick-forklift-example/blob/master/migrations/src/main/scala/Migrations.scala) to tell Forklift where to read the database configurations).

I wonder if this project fits your requirements? Please let me know if you have any questions.

**Comment by krivachy** [comment]
@lastland thanks for the effort, I checked out the example project and it works great :+1: It would be something I could start using in a project in the coming weeks and give you more feedback.

Where can I find out more info about when I would use `datamodel.v1.schema`? I don't quite see the point in ever referencing an older schema from my code.

Thanks again for the clear example and Scala Forklift is very cool!

**Comment by cvogt** [comment]
migrations change the schema. if they are type-safe, they need to type-check against the version they upgrade. If you upgrade with multiple migrations, each one needs a different schema that corresponds to the database at that point in time. That's why you need to keep the older versions around, to upgrade older database versions. If that is never the case they could be thrown away after the upgrade.

**Comment by cvogt** [comment]
And I agree, Forklift is very cool ;)

**Comment by lastland** [comment]
@krivachy Glad you find it useful! :smile: 

The reason to keep `datamodel.v1.schema` is exactly as @cvogt has explained.

Please also note that I have just published Forklift version `0.2.1` (which supports Slick 3.1.1) and `0.2.0` (which supports Slick 3.0). I have also updated the starting template to show how to use Forklift 0.2.1 with slick 3.1.1. See: https://github.com/lastland/scala-forklift-start-template
